### DISCUSSION TITLE
Import builtin module without prefix

### DISCUSSION THREAD

**Discussion by MaxGyver83** [janet-lang/janet#1388] [Q&A] [answered]
I would like to import `(os)` without prefix:

```janet
(import os :prefix "")
```

This fails because this module can't be found.

I'm aware that `os` is already imported but with a prefix. How can I import it without prefix?

**Comment by iacore** [comment, marked as answer]
You want to assign new variables. Here's how you get the names with "os/".

```
(->> (all-bindings) (filter |(string/has-prefix? "os/" $)))
```

**Reply by MaxGyver83** [reply]
Thank you for this hint. I have tried to assign new variables like this:

```janet
(->> (all-bindings) (filter |(string/has-prefix? "os/" $)) |(def (string/replace "os/" "" $) $))
```

This doesn't work. The added forms is like:

```janet
(def "cd" "os/cd")
```

But I want to do

```janet
(def cd os/cd)
```

How can I convert a string to a symbol?

**Reply by MaxGyver83** [reply]
I have also tried this:

```janet
(->> (all-bindings)
  (filter |(string/has-prefix? "os/" $))
  |(eval-string (string "(def " (string/replace "os/" "" $) " " $ ")")))
```

This gives me:

```
compile error: unexpected type in destruction, got "os/"
```

**Reply by MaxGyver83** [reply]
OK, I forgot to add "map". Now it works:

```janet
(->> (all-bindings)
     (filter |(string/has-prefix? "os/" $))
     (map |(eval-string (string "(def " (string/replace "os/" "" $) " " $ ")"))))
```

Thank you!

**Reply by ianthehenry** [reply]
There's another way to do what you want without using `eval`.

```janet
(loop [[name entry] :pairs root-env
       :when (string/has-prefix? "os/" name)]
  (def unprefixed-name (symbol/slice name 3))
  (def local-entry (table/setproto @{:private true} entry))
  (put (curenv) unprefixed-name local-entry))
```

Environments are tables, and to modify environments you can use `put`. No need for `eval`!

The `local-entry` thing is kind of tricky; this will alias the names as private identifiers so that if someone imports *this* module, they won't get names like `my-module/spawn` and such. This is the behavior if you use `def-` instead of `def`.

`eval` is a little less code, though. I'd note that it's easier to construct an s-expression than to use `eval-string`:

```janet
(eval-string (string "(def " (string/replace "os/" "" $) " " $ ")"))
(eval ~(def- ,(string/replace "os/" "" $) ,$))
```

**Reply by iacore** [reply]
You can set the current environment directly.

`(set ((curenv) 'cd) (dyn 'os/cd))`

**Reply by ianthehenry** [reply]
You can use `set` to mutate the environment, but `put` does the same thing with fewer parens.  ¯\\_(ツ)\_/¯

**Reply by sogaiu** [reply]
Sorry, couldn't resist the [temptation](http://www.paren.party/).

**Reply by MaxGyver83** [reply]
> ```janet
> (eval-string (string "(def " (string/replace "os/" "" $) " " $ ")"))
> (eval ~(def- ,(string/replace "os/" "" $) ,$))
> ```

This doesn't work:

```janet
repl:1:> (->> (all-bindings)
repl:2:(>      (filter |(string/has-prefix? "os/" $))
repl:3:(>      (map |(eval ~(def- ,(string/replace "os/" "" $) ,$))))
error: unexpected type in destruction, got "arch"
  in eval [boot.janet] (tailcall) on line 2665, column 5
  in map [boot.janet] on line 998, column 3
  in _thunk [repl] (tailcall) on line 3, column 6
```

Do you know what's wrong here?

**Reply by ianthehenry** [reply]
Oh, yeah, sorry -- the `string/replace` implicitly converts your symbol into a string and returns a string, but we need a symbol there. Using `,(symbol/slice $ 3)` instead will work.

This type difference is lost when you turn the whole thing into a string -- strings and symbols have the same string representation -- but it's important in the s-expression. It's the difference between typing `(def "arch" os/arch)` and `(def arch os/arch)` at the repl -- the former is not allowed.
